google/cloud/internal/grpc_opentelemetry.cc

```diff
@@ -0,0 +1,52 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/grpc_opentelemetry.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/options.h"
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include <opentelemetry/trace/semantic_conventions.h>
+#include <opentelemetry/trace/span_metadata.h>
+#include <opentelemetry/trace/span_startoptions.h>
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeSpanGrpc(
+    opentelemetry::nostd::string_view service,
+    opentelemetry::nostd::string_view method) {
+  namespace SC = opentelemetry::trace::SemanticConventions;
+  opentelemetry::trace::StartSpanOptions options;
+  options.kind = opentelemetry::trace::SpanKind::kClient;
+  return GetTracer(internal::CurrentOptions())
+      ->StartSpan(absl::StrCat(service.data(), "/", method.data()),
+                  {{SC::kRpcSystem, SC::RpcSystemValues::kGrpc},
+                   {SC::kRpcService, service},
+                   {SC::kRpcMethod, method},
+                   {SC::kNetTransport, SC::NetTransportValues::kIpTcp}},
+                  options);
```

Comment:
Is there a place to capture the `google-cloud-cpp` version and the `grpc` library versions?

Reply:
Even if there is no standard place: should we capture them to help our troubleshooting efforts?

Reply:
> capture the google-cloud-cpp version?

The version of our library is captured in the instrumentation scope: https://github.com/googleapis/google-cloud-cpp/blob/70543f8de9edaea21e98370527f280f2ac8ee95f/google/cloud/internal/opentelemetry.cc#L34

The awkward thing is that the Cloud Trace `Span` message [does not have a dedicated field for this info](https://github.com/googleapis/googleapis/blob/4eb9a0e60227e1bbe8b80eb9553f463e43ba1817/google/devtools/cloudtrace/v2/trace.proto#L219-L277). But that is a problem for the exporter, which will likely set the instrumentation scope as attributes.

> capture the grpc library version?

I will address this in another commit...

Reply:
SG

Reply:
Added `"grpc.version"` attribute.